In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Load the Database using sqlite

In [25]:
import sqlite3
import pandas as pd

# 1. Create a connection to the database
conn = sqlite3.connect('/content/drive/MyDrive/Y3 S1/AI solution dev pj/Project/data/gas_monitoring.db')

# 2. Query to find all the table names in the database
cursor = conn.cursor()
cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = cursor.fetchall()

print("Tables in the database:", tables)

Tables in the database: [('gas_monitoring',)]


# Extract the Data into a DataFrame

In [26]:
# Replace 'sensor_data' with the actual table name you found in Step 1
query = "SELECT * FROM gas_monitoring"

# Load the data into a Pandas DataFrame
df = pd.read_sql_query(query, conn)

# Close the connection since we have the data in memory now
conn.close()

# Look at the first 5 rows to see what you're dealing with
display(df.head())

,Time of Day,Temperature,Humidity,CO2_InfraredSensor,CO2_ElectroChemicalSensor,MetalOxideSensor_Unit1,MetalOxideSensor_Unit2,MetalOxideSensor_Unit3,MetalOxideSensor_Unit4,CO_GasSensor,Session ID,HVAC Operation Mode,Ambient Light Level,Activity Level
0,morning,19.63,53.83,125.486389,571.089440,478.554958,735.850412,654.625253,565.051969,2.0,7241,off,very_dim,Low Activity
1,morning,292.81,53.69,126.343018,575.789501,491.955151,740.282738,655.734327,557.078486,1.0,7241,ventilation_only,bright,Low Activity
2,morning,19.62,54.25,126.560695,585.543111,505.560808,737.112906,649.962421,558.065196,1.0,7241,eco_mode,None,Low Activity
3,morning,20.10,50.48,113.504877,597.449961,515.142272,744.020651,676.150075,600.222210,NaN,7241,eco_mode,very_bright,High Activity
4,morning,19.89,48.42,92.766225,613.654875,535.664558,770.265441,720.274019,625.731124,1.0,7241,heating_active,moderate,LowActivity


# The "Sanity Check"

Checking the missing values in the dataframe

After checking, we found out that there are many missing values, impossible outliers like having 300 degrees in temperature and -49% in humidity.

In [27]:
# Check the shape (number of rows and columns)
print(f"Dataset contains {df.shape[0]} rows and {df.shape[1]} columns.\n")

# Check data types and look for missing (null) values
print("--- Data Info ---")
df.info()

# Get a statistical summary to spot obvious outliers
# (e.g., a room temperature of 999 degrees)
display(df.describe())

Dataset contains 10000 rows and 14 columns.

--- Data Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Time of Day                10000 non-null  object 
 1   Temperature                10000 non-null  float64
 2   Humidity                   8072 non-null   float64
 3   CO2_InfraredSensor         10000 non-null  float64
 4   CO2_ElectroChemicalSensor  10000 non-null  float64
 5   MetalOxideSensor_Unit1     10000 non-null  float64
 6   MetalOxideSensor_Unit2     8590 non-null   float64
 7   MetalOxideSensor_Unit3     10000 non-null  float64
 8   MetalOxideSensor_Unit4     10000 non-null  float64
 9   CO_GasSensor               9166 non-null   float64
 10  Session ID                 10000 non-null  int64  
 11  HVAC Operation Mode        10000 non-null  object 
 12  Ambient Light Level        8946 non-null

,Temperature,Humidity,CO2_InfraredSensor,CO2_ElectroChemicalSensor,MetalOxideSensor_Unit1,MetalOxideSensor_Unit2,MetalOxideSensor_Unit3,MetalOxideSensor_Unit4,CO_GasSensor,Session ID
count,10000.000000,8072.000000,10000.000000,10000.000000,10000.000000,8590.000000,10000.000000,10000.000000,9166.000000,10000.000000
mean,42.243124,51.643609,109.864352,578.634082,470.932316,728.354408,680.488646,612.247715,1.288239,5427.921400
std,73.974956,21.333894,33.793851,22.573537,51.459025,29.140123,55.953530,43.174354,0.779754,2590.836924
min,6.903485,-49.870000,-17.472184,408.599386,286.825662,611.168916,456.673900,412.791904,0.000000,1374.000000
25%,18.937500,47.660000,96.660377,559.353100,440.035413,709.660122,658.519775,585.511731,1.000000,3074.000000
50%,20.510000,51.390000,112.807103,579.361082,469.146955,726.748925,679.908353,609.526245,1.000000,5214.000000
75%,23.172500,53.840000,124.540226,595.569759,494.770045,744.879789,703.522845,636.012105,2.000000,7395.000000
max,307.070000,198.900000,237.873938,637.895790,632.891693,807.732943,906.213097,736.112361,4.000000,9658.000000


# Data Cleaning

Remove extreme outliers but it will also remove the null since the method is to remove everything which is not in the range.

In [28]:
# Document assumption: Negative gases/light and extreme temps are sensor errors.
df_filtered = df[
    (df['Temperature'] >= 10) & (df['Temperature'] <= 40) &
    (df['Humidity'] >= 0) & (df['Humidity'] <= 100) &
    (df['CO2_InfraredSensor'] >= 0) & (df['CO2_InfraredSensor'] <= 10000) &
    (df['CO2_ElectroChemicalSensor'] >= 0) & (df['CO2_ElectroChemicalSensor'] <= 10000) &
    (df['CO_GasSensor'] >= 0) & (df['CO_GasSensor'] <= 2000) &
    (df['MetalOxideSensor_Unit1'] >= 0) &
    (df['MetalOxideSensor_Unit2'] >= 0) &
    (df['MetalOxideSensor_Unit3'] >= 0) &
    (df['MetalOxideSensor_Unit4'] >= 0)
].copy()

# Filtering for Metal Oxide Sensors (Removing the top 1% as glitches)
# We calculate the 99th percentile and keep only data below it
for col in ['MetalOxideSensor_Unit1', 'MetalOxideSensor_Unit2', 'MetalOxideSensor_Unit3', 'MetalOxideSensor_Unit4']:
    cap_value = df_filtered[col].quantile(0.99)
    df_filtered = df_filtered[df_filtered[col] <= cap_value]

print(f"Original rows: {len(df)}")
print(f"Rows after comprehensive outlier removal: {len(df_filtered)}")

Original rows: 10000
Rows after comprehensive outlier removal: 5238


Filling Null (Missing) Values with the mode

In [29]:
#Check for missing values before imputation
print(df_filtered.isnull().sum())

Time of Day                    0
Temperature                    0
Humidity                       0
CO2_InfraredSensor             0
CO2_ElectroChemicalSensor      0
MetalOxideSensor_Unit1         0
MetalOxideSensor_Unit2         0
MetalOxideSensor_Unit3         0
MetalOxideSensor_Unit4         0
CO_GasSensor                   0
Session ID                     0
HVAC Operation Mode            0
Ambient Light Level          570
Activity Level                 0
dtype: int64


Do .info to check the number is correct or not

In [30]:
df_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5238 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Time of Day                5238 non-null   object 
 1   Temperature                5238 non-null   float64
 2   Humidity                   5238 non-null   float64
 3   CO2_InfraredSensor         5238 non-null   float64
 4   CO2_ElectroChemicalSensor  5238 non-null   float64
 5   MetalOxideSensor_Unit1     5238 non-null   float64
 6   MetalOxideSensor_Unit2     5238 non-null   float64
 7   MetalOxideSensor_Unit3     5238 non-null   float64
 8   MetalOxideSensor_Unit4     5238 non-null   float64
 9   CO_GasSensor               5238 non-null   float64
 10  Session ID                 5238 non-null   int64  
 11  HVAC Operation Mode        5238 non-null   object 
 12  Ambient Light Level        4668 non-null   object 
 13  Activity Level             5238 non-null   object 
dt

Since Ambient Light Level is missing 570 rows so I filled with the mode which is most frequently appeared in the database.

In [31]:
# Calculate the mode (most common category) for Ambient Light Level
light_mode = df_filtered['Ambient Light Level'].mode()[0]

# Fill the missing values with the mode
df_filtered['Ambient Light Level'] = df_filtered['Ambient Light Level'].fillna(light_mode)

# Final Sanity Check to ensure everything is at 0
print("Missing values after filling Ambient Light Level:")
print(df_filtered.isnull().sum())

Missing values after filling Ambient Light Level:
Time of Day                  0
Temperature                  0
Humidity                     0
CO2_InfraredSensor           0
CO2_ElectroChemicalSensor    0
MetalOxideSensor_Unit1       0
MetalOxideSensor_Unit2       0
MetalOxideSensor_Unit3       0
MetalOxideSensor_Unit4       0
CO_GasSensor                 0
Session ID                   0
HVAC Operation Mode          0
Ambient Light Level          0
Activity Level               0
dtype: int64


In [32]:
df_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5238 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Time of Day                5238 non-null   object 
 1   Temperature                5238 non-null   float64
 2   Humidity                   5238 non-null   float64
 3   CO2_InfraredSensor         5238 non-null   float64
 4   CO2_ElectroChemicalSensor  5238 non-null   float64
 5   MetalOxideSensor_Unit1     5238 non-null   float64
 6   MetalOxideSensor_Unit2     5238 non-null   float64
 7   MetalOxideSensor_Unit3     5238 non-null   float64
 8   MetalOxideSensor_Unit4     5238 non-null   float64
 9   CO_GasSensor               5238 non-null   float64
 10  Session ID                 5238 non-null   int64  
 11  HVAC Operation Mode        5238 non-null   object 
 12  Ambient Light Level        5238 non-null   object 
 13  Activity Level             5238 non-null   object 
dt